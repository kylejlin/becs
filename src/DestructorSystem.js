class DestructorSystem {
  constructor(componentNames, destroy) {
    this.componentNames = componentNames;
    this.destroy = destroy;
  }
}

export default DestructorSystem;
