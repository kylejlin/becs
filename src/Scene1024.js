import Scene from './Scene';

class Scene1024 extends Scene {
  static maxComponentTypes_ = 1024;

  static equals_(a, b) {
    return a[0] === b[0]
      && a[1] === b[1]
      && a[2] === b[2]
      && a[3] === b[3]
      && a[4] === b[4]
      && a[5] === b[5]
      && a[6] === b[6]
      && a[7] === b[7]
      && a[8] === b[8]
      && a[9] === b[9]
      && a[10] === b[10]
      && a[11] === b[11]
      && a[12] === b[12]
      && a[13] === b[13]
      && a[14] === b[14]
      && a[15] === b[15]
      && a[16] === b[16]
      && a[17] === b[17]
      && a[18] === b[18]
      && a[19] === b[19]
      && a[20] === b[20]
      && a[21] === b[21]
      && a[22] === b[22]
      && a[23] === b[23]
      && a[24] === b[24]
      && a[25] === b[25]
      && a[26] === b[26]
      && a[27] === b[27]
      && a[28] === b[28]
      && a[29] === b[29]
      && a[30] === b[30]
      && a[31] === b[31];
  }

  // Checks if a âŠ† b
  static subset_(a, b) {
    return (~a[0] | b[0]) === -1
      && (~a[1] | b[1]) === -1
      && (~a[2] | b[2]) === -1
      && (~a[3] | b[3]) === -1
      && (~a[4] | b[4]) === -1
      && (~a[5] | b[5]) === -1
      && (~a[6] | b[6]) === -1
      && (~a[7] | b[7]) === -1
      && (~a[8] | b[8]) === -1
      && (~a[9] | b[9]) === -1
      && (~a[10] | b[10]) === -1
      && (~a[11] | b[11]) === -1
      && (~a[12] | b[12]) === -1
      && (~a[13] | b[13]) === -1
      && (~a[14] | b[14]) === -1
      && (~a[15] | b[15]) === -1
      && (~a[16] | b[16]) === -1
      && (~a[17] | b[17]) === -1
      && (~a[18] | b[18]) === -1
      && (~a[19] | b[19]) === -1
      && (~a[20] | b[20]) === -1
      && (~a[21] | b[21]) === -1
      && (~a[22] | b[22]) === -1
      && (~a[23] | b[23]) === -1
      && (~a[24] | b[24]) === -1
      && (~a[25] | b[25]) === -1
      && (~a[26] | b[26]) === -1
      && (~a[27] | b[27]) === -1
      && (~a[28] | b[28]) === -1
      && (~a[29] | b[29]) === -1
      && (~a[30] | b[30]) === -1
      && (~a[31] | b[31]) === -1;
  }
}

export default Scene1024;
