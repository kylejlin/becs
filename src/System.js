class System {
  constructor(filters, update) {
    this.filters = filters;
    this.update = update;
  }
}

export default System;
