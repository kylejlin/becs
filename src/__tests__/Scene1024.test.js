import Scene1024 from '../Scene1024';
import System from '../System';
import DestructorSystem from '../DestructorSystem';

test('systems work correctly', () => {
  const scene = new Scene1024();
  const [bob, joe, larry] = [
    {
      Position: { x: 0, y: 0, z: 0 },
      Name: { name: 'Bob' },
    },
    {
      Position: { x: 0, y: 0, z: 0 },
      Velocity: { x: 1, y: 0, z: 0 },
      Name: { name: 'Joe' },
    },
    {
      Position: { x: 0, y: 10, z: 0 },
      Velocity: { x: 0, y: 5, z: 8 },
      Name: { name: 'Larry' },
    }
  ];
  scene.addEntity(bob);
  scene.addEntity(joe);
  const inertia = new System(
    ['Position', 'Velocity'],
    (entities) => {
      for (const ent of entities) {
        ent.Position.x += ent.Velocity.x;
        ent.Position.y += ent.Velocity.y;
        ent.Position.z += ent.Velocity.z;
      }
    }
  );
  scene.addSystem(inertia);
  scene.addEntity(larry);
  expect(scene).toMatchSnapshot({}, 'initial scene');
  scene.update();
  expect(scene).toMatchSnapshot({}, 'with inertia added');
  scene.removeEntity(joe);
  scene.update();
  expect(scene).toMatchSnapshot({}, 'with joe removed');
  scene.removeSystem(inertia);
  scene.update();
  expect(scene).toMatchSnapshot({}, 'with inertia removed');
});

test('destructors work correctly', () => {
  let freed = null;

  const scene = new Scene1024();
  const [bob, joe, larry] = [
    {
      Position: { x: 0, y: 0, z: 0 },
      Name: { name: 'Bob' },
    },
    {
      Position: { x: 0, y: 0, z: 0 },
      Velocity: { x: 1, y: 0, z: 0 },
      Name: { name: 'Joe' },
    },
    {
      Position: { x: 0, y: 10, z: 0 },
      Velocity: { x: 0, y: 5, z: 8 },
      Name: { name: 'Larry' },
    }
  ];
  scene.addEntity(bob);
  scene.addEntity(joe);
  const freer = new DestructorSystem(
    ['Position', 'Velocity', 'Name'],
    (entity) => {
      freed = entity.Name.name;
    }
  );
  scene.addDestructorSystem(freer);
  scene.removeEntity(joe);
  expect(freed).toEqual('Joe');
  freed = null;
  scene.removeEntity(bob);
  expect(freed).toEqual(null);
  scene.addEntity(larry);
  scene.removeEntity(larry);
  expect(freed).toEqual('Larry');
});
