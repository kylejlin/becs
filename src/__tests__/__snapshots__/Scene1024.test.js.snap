// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`systems work correctly: initial scene 1`] = `
Scene1024 {
  "componentNumberDict_": Object {
    "Name": 1,
    "Position": 0,
    "Velocity": 2,
  },
  "destructorSystems_": Array [],
  "entities_": Array [
    Object {
      "Name": Object {
        "name": "Bob",
      },
      "Position": Object {
        "x": 0,
        "y": 0,
        "z": 0,
      },
    },
    Object {
      "Name": Object {
        "name": "Joe",
      },
      "Position": Object {
        "x": 0,
        "y": 0,
        "z": 0,
      },
      "Velocity": Object {
        "x": 1,
        "y": 0,
        "z": 0,
      },
    },
    Object {
      "Name": Object {
        "name": "Larry",
      },
      "Position": Object {
        "x": 0,
        "y": 10,
        "z": 0,
      },
      "Velocity": Object {
        "x": 0,
        "y": 5,
        "z": 8,
      },
    },
  ],
  "globals": Object {},
  "indexes_": Array [
    Object {
      "bitmask": Int32Array [
        5,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "dependents": 1,
      "entities": Array [
        Object {
          "Name": Object {
            "name": "Joe",
          },
          "Position": Object {
            "x": 0,
            "y": 0,
            "z": 0,
          },
          "Velocity": Object {
            "x": 1,
            "y": 0,
            "z": 0,
          },
        },
        Object {
          "Name": Object {
            "name": "Larry",
          },
          "Position": Object {
            "x": 0,
            "y": 10,
            "z": 0,
          },
          "Velocity": Object {
            "x": 0,
            "y": 5,
            "z": 8,
          },
        },
      ],
    },
  ],
  "numberOfRegisteredComponents_": 3,
  "systems_": Array [
    Object {
      "index": Object {
        "bitmask": Int32Array [
          5,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "dependents": 1,
        "entities": Array [
          Object {
            "Name": Object {
              "name": "Joe",
            },
            "Position": Object {
              "x": 0,
              "y": 0,
              "z": 0,
            },
            "Velocity": Object {
              "x": 1,
              "y": 0,
              "z": 0,
            },
          },
          Object {
            "Name": Object {
              "name": "Larry",
            },
            "Position": Object {
              "x": 0,
              "y": 10,
              "z": 0,
            },
            "Velocity": Object {
              "x": 0,
              "y": 5,
              "z": 8,
            },
          },
        ],
      },
      "rawRef": System {
        "componentNames": Array [
          "Position",
          "Velocity",
        ],
        "update": [Function],
      },
      "update": [Function],
    },
  ],
}
`;

exports[`systems work correctly: with inertia added 1`] = `
Scene1024 {
  "componentNumberDict_": Object {
    "Name": 1,
    "Position": 0,
    "Velocity": 2,
  },
  "destructorSystems_": Array [],
  "entities_": Array [
    Object {
      "Name": Object {
        "name": "Bob",
      },
      "Position": Object {
        "x": 0,
        "y": 0,
        "z": 0,
      },
    },
    Object {
      "Name": Object {
        "name": "Joe",
      },
      "Position": Object {
        "x": 1,
        "y": 0,
        "z": 0,
      },
      "Velocity": Object {
        "x": 1,
        "y": 0,
        "z": 0,
      },
    },
    Object {
      "Name": Object {
        "name": "Larry",
      },
      "Position": Object {
        "x": 0,
        "y": 15,
        "z": 8,
      },
      "Velocity": Object {
        "x": 0,
        "y": 5,
        "z": 8,
      },
    },
  ],
  "globals": Object {},
  "indexes_": Array [
    Object {
      "bitmask": Int32Array [
        5,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "dependents": 1,
      "entities": Array [
        Object {
          "Name": Object {
            "name": "Joe",
          },
          "Position": Object {
            "x": 1,
            "y": 0,
            "z": 0,
          },
          "Velocity": Object {
            "x": 1,
            "y": 0,
            "z": 0,
          },
        },
        Object {
          "Name": Object {
            "name": "Larry",
          },
          "Position": Object {
            "x": 0,
            "y": 15,
            "z": 8,
          },
          "Velocity": Object {
            "x": 0,
            "y": 5,
            "z": 8,
          },
        },
      ],
    },
  ],
  "numberOfRegisteredComponents_": 3,
  "systems_": Array [
    Object {
      "index": Object {
        "bitmask": Int32Array [
          5,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "dependents": 1,
        "entities": Array [
          Object {
            "Name": Object {
              "name": "Joe",
            },
            "Position": Object {
              "x": 1,
              "y": 0,
              "z": 0,
            },
            "Velocity": Object {
              "x": 1,
              "y": 0,
              "z": 0,
            },
          },
          Object {
            "Name": Object {
              "name": "Larry",
            },
            "Position": Object {
              "x": 0,
              "y": 15,
              "z": 8,
            },
            "Velocity": Object {
              "x": 0,
              "y": 5,
              "z": 8,
            },
          },
        ],
      },
      "rawRef": System {
        "componentNames": Array [
          "Position",
          "Velocity",
        ],
        "update": [Function],
      },
      "update": [Function],
    },
  ],
}
`;

exports[`systems work correctly: with inertia removed 1`] = `
Scene1024 {
  "componentNumberDict_": Object {
    "Name": 1,
    "Position": 0,
    "Velocity": 2,
  },
  "destructorSystems_": Array [],
  "entities_": Array [
    Object {
      "Name": Object {
        "name": "Bob",
      },
      "Position": Object {
        "x": 0,
        "y": 0,
        "z": 0,
      },
    },
    Object {
      "Name": Object {
        "name": "Larry",
      },
      "Position": Object {
        "x": 0,
        "y": 20,
        "z": 16,
      },
      "Velocity": Object {
        "x": 0,
        "y": 5,
        "z": 8,
      },
    },
  ],
  "globals": Object {},
  "indexes_": Array [],
  "numberOfRegisteredComponents_": 3,
  "systems_": Array [],
}
`;

exports[`systems work correctly: with joe removed 1`] = `
Scene1024 {
  "componentNumberDict_": Object {
    "Name": 1,
    "Position": 0,
    "Velocity": 2,
  },
  "destructorSystems_": Array [],
  "entities_": Array [
    Object {
      "Name": Object {
        "name": "Bob",
      },
      "Position": Object {
        "x": 0,
        "y": 0,
        "z": 0,
      },
    },
    Object {
      "Name": Object {
        "name": "Larry",
      },
      "Position": Object {
        "x": 0,
        "y": 20,
        "z": 16,
      },
      "Velocity": Object {
        "x": 0,
        "y": 5,
        "z": 8,
      },
    },
  ],
  "globals": Object {},
  "indexes_": Array [
    Object {
      "bitmask": Int32Array [
        5,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
      "dependents": 1,
      "entities": Array [
        Object {
          "Name": Object {
            "name": "Larry",
          },
          "Position": Object {
            "x": 0,
            "y": 20,
            "z": 16,
          },
          "Velocity": Object {
            "x": 0,
            "y": 5,
            "z": 8,
          },
        },
      ],
    },
  ],
  "numberOfRegisteredComponents_": 3,
  "systems_": Array [
    Object {
      "index": Object {
        "bitmask": Int32Array [
          5,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
        ],
        "dependents": 1,
        "entities": Array [
          Object {
            "Name": Object {
              "name": "Larry",
            },
            "Position": Object {
              "x": 0,
              "y": 20,
              "z": 16,
            },
            "Velocity": Object {
              "x": 0,
              "y": 5,
              "z": 8,
            },
          },
        ],
      },
      "rawRef": System {
        "componentNames": Array [
          "Position",
          "Velocity",
        ],
        "update": [Function],
      },
      "update": [Function],
    },
  ],
}
`;
